runtime/sema/check_swap.go

```diff
@@ -25,38 +25,28 @@ import (
 
 func (checker *Checker) VisitSwapStatement(swap *ast.SwapStatement) (_ struct{}) {
 
-	leftType := checker.VisitExpression(swap.Left, nil)
-	rightType := checker.VisitExpression(swap.Right, nil)
+	// First visit the two expressions as if they were the target of the assignment.
+	leftTargetType := checker.checkSwapStatementExpression(swap.Left, common.OperandSideLeft)
+	rightTargetType := checker.checkSwapStatementExpression(swap.Right, common.OperandSideRight)
+
+	// Then re-visit the same expressions, this time treat them as the value-expr of the assignment.
+	// The 'expected type' of the two expression would be the types obtained from the previous visit, swapped.
+	leftValueType := checker.VisitExpression(swap.Left, rightTargetType)
+	rightValueType := checker.VisitExpression(swap.Right, leftTargetType)
```

Comment:
This change is more like a clean-up: But also needed to handle when the swap-statement has references.